### DISCUSSION TITLE
how does the Xcode find its version

### DISCUSSION THREAD

**Discussion by the-code-samurai-97** [bazelbuild/rules_apple#2348] [Q&A] [answered]
```
 ERROR: /private/var/tmp/_bazel_sibinarayananrevathi/7323e08054f8da7553e60a0a337f01b0/external/local_config_apple_cc/BUILD:79:24: in cc_toolchain_config rule @local_config_apple_cc//:darwin_arm64: 
Traceback (most recent call last):
        File "/private/var/tmp/_bazel_sibinarayananrevathi/7323e08054f8da7553e60a0a337f01b0/external/local_config_apple_cc/cc_toolchain_config.bzl", line 2465, column 58, in _impl
                enabled = xcode_support.is_xcode_at_least_version(xcode_config, "15.0.0"),
        File "/private/var/tmp/_bazel_sibinarayananrevathi/7323e08054f8da7553e60a0a337f01b0/external/build_bazel_apple_support/lib/xcode_support.bzl", line 35, column 13, in _is_xcode_at_least_version
                fail("Could not determine Xcode version at all. This likely means Xcode isn't available; " +
Error in fail: Could not determine Xcode version at all. This likely means Xcode isn't available; if you think this is a mistake, please file an issue.
(13:35:27) ERROR: /private/var/tmp/_bazel_sibinarayananrevathi/7323e08054f8da7553e60a0a337f01b0/external/local_config_apple_cc/BUILD:79:24: Analysis of target '@local_config_apple_cc//:darwin_arm64' failed

```
xcode-select --version is working
what am i miising here?

**Comment by keith** [comment, marked as answer]
Are you using a standard recent macOS installation? (I ask because of your gcc thread)

Bazel uses spotlight APIs to discover Xcode versions, this likely indicates it couldn't find any. You can try to run:

```
cat $(bazel info output_base)/external/*local_config_xcode/BUILD
```

To see what is in the generated BUILD file that should have all the discovered versions (I'm not sure if this works in this case because it might fail too early)

**Reply by the-code-samurai-97** [reply]
package(default_visibility = ['//visibility:public'])

xcode_version(
  name = 'version15_1_0_15C65',
  version = '15.1.0.15C65',
  aliases = ['15.1', '15.1.0', '15', '15C65', '15.1.0.15C65'],
  default_ios_sdk_version = '17.2',
  default_tvos_sdk_version = '17.2',
  default_macos_sdk_version = '14.2',
  default_watchos_sdk_version = '10.2',
)
xcode_config(
  name = 'host_xcodes',
  versions = [':version15_1_0_15C65'],
  default = ':version15_1_0_15C65',
)
available_xcodes(
  name = 'host_available_xcodes',
  versions = [':version15_1_0_15C65'],
  default = ':version15_1_0_15C65',
)

**Reply by keith** [reply]
that looks right, you can try forcing the version with `--xcode_version=15.1` (shouldn't be required but maybe that will work). If you installed Xcode after you had run bazel you might need to do a `bazel shutdown` and try again. Alternatively you can try this method to workaround https://www.smileykeith.com/2021/03/08/locking-xcode-in-bazel/

**Comment by the-code-samurai-97** [comment]
Thanks @keith I will check it out ! I encountered the error while compiling this Repo if you have time to take a look ,https://github.com/mvukov/rules_ros2 I was using clang-14 which the author said it worked while compiling I faced this, i am new to Bazel and mostly I use Linux for work and for personal I got this Mac errors

**Comment by kennethvl** [comment]
Hi, I also encountered a similar issue. I used [xcodes](https://github.com/XcodesOrg/xcodes) to install the Xcode. I've confirmed the the xcode-select path is there:

```
% xcode-select -p
/Applications/Xcode-15.2.0.app/Contents/Developer
```

But bazel failed to detect the installed Xcode:

```
% cat local_config_xcode/BUILD
package(default_visibility = ['//visibility:public'])

xcode_config(name = 'host_xcodes')
# Error: Invoking xcode-locator failed, return code 1, stderr: error: Error Domain=NSOSStatusErrorDomain Code=-10814 "kLSApplicationNotFoundErr: E.g. no application claims the file" UserInfo={_LSLine=1569, _LSFunction=runEvaluator} , stdout:
```

Did I miss something during my setup?

**Reply by brentleyjones** [reply]
We debugged the issue in slack: https://bazelbuild.slack.com/archives/CD3QY5C2X/p1709523435623459?thread_ts=1709022691.178389&channel=CD3QY5C2X&message_ts=1709523435.623459

**Reply by sfc-gh-kleonhard** [reply]
@brentleyjones - was the debugging fruitful?

**Reply by brentleyjones** [reply]
Yes. Their macOS version didn't support the selected Xcode version.

**Reply by sfc-gh-kleonhard** [reply]
@brentleyjones can I trouble you to summarize the debug steps? I don't seem to have access to the slack workspace or I'd follow along directly.

I experience similar symptoms on upgrade to Bazel 7.1.1 from Bazel 6.5.0.

Setting `--repo_env=BAZEL_NO_APPLE_CPP_TOOLCHAIN=1` avoids the error but causes a python build to fail on `Undefined symbols for architecture arm64`. Interestingly, the same python build on Bazel 6.5.0 with BAZEL_NO_APPLE_CPP_TOOLCHAIN=1 set works.

```
cat $(bazel info output_base)/external/*local_config_xcode/BUILD

package(default_visibility = ['//visibility:public'])

xcode_config(name = 'host_xcodes')
# Error: Invoking xcode-locator failed, return code 1, stderr: error: Error Domain=NSOSStatusErrorDomain Code=-10814 "kLSApplicationNotFoundErr: E.g. no application claims the file" UserInfo={_LSLine=1741, _LSFunction=runEvaluator} , stdout: 
```

```
xcode-select -p
/Library/Developer/CommandLineTools
```

**Comment by proppy** [comment]
https://stackoverflow.com/a/46460129 this worked for me.

**Comment by cshroba-figma** [comment]
Using the App Store to update Xcode to the newest version solved this problem for me after I updated macOS.
